src/Fmt_ast.ml

```diff
@@ -3635,15 +3635,47 @@ and fmt_structure c ctx itms =
           | _ -> false
         in
         let is_simple itm =
-          match itm.pstr_desc with
-          | Pstr_include {pincl_mod= me} | Pstr_module {pmb_expr= me} ->
+          match (c.conf.let_binding, itm.pstr_desc) with
+          | _, Pstr_include {pincl_mod= me} | _, Pstr_module {pmb_expr= me} ->
               is_simple_mod me
-          | Pstr_open _ -> true
+          | _, Pstr_open _ -> true
+          | `Compact, _ ->
+              let width i =
+                let last = false in
+                String.length
+                  (Format.asprintf "%t"
+                     (Cmts.preserve (fun () -> fmt_structure_item ~last c i)))
+              in
+              width (sub_str ~ctx itm) <= c.conf.margin
           | _ -> false
         in
+        let group itmI itmJ =
+          match (c.conf.let_binding, itmI.pstr_desc, itmJ.pstr_desc) with
```

Comment:
I modified this function (formerly `break_between_open_or_include`), is this what you had in mind ? (not much shorter)

Reply:
Yes, basically. Some of the repetition of Compact could be avoided with a nested or-pattern, but not a big deal.

Regarding the actual grouping, before this patch simple `open` and `module` items are formatted without an open line between, which might be good to preserve. I am not sure what the best grouping is though. In practice my feeling is that `open`, `include`, and `module` (at least for cases like `module A = B`) are often conceptually good to group together. But it may take some experimentation.